h2o-persist-hdfs/src/main/java/water/persist/security/HdfsDelegationTokenRefresher.java

```diff
@@ -0,0 +1,230 @@
+        _retryDelaySecs = conf.getInt(H2O_AUTH_TOKEN_REFRESHER_RETRY_DELAY_SECS, 10);
+        _fallbackIntervalSecs = conf.getInt(H2O_AUTH_TOKEN_REFRESHER_FALLBACK_INTERVAL_SECS, 12 * 3600); // 12h
+    }
+
+    public void start() {
+        long intervalSecs = 0L;
+        try {
+             intervalSecs = getTokenRenewalIntervalSecs(loginAuthUser());
+        } catch (IOException | InterruptedException e) {
+            log("Encountered error while trying to determine token renewal interval.", e);
+        }
+        final long actualIntervalSecs;
+        if (intervalSecs == 0L) {
+            actualIntervalSecs = _fallbackIntervalSecs;
+            log("Token renewal interval was not determined, will use " + _fallbackIntervalSecs + "s.", null);
+        } else {
+            actualIntervalSecs = (long) (intervalSecs * _intervalRatio);
+            log("Determined token renewal interval = " + intervalSecs + "s. " +
+                    "Using actual interval = " + actualIntervalSecs + "s (ratio=" + _intervalRatio + ").", null);
+        }
+        _executor.scheduleAtFixedRate(this, 0, actualIntervalSecs, TimeUnit.SECONDS);
+    }
+
+    private static void log(String s, Exception e) {
+        System.out.println("HDFS TOKEN REFRESH: " + s);
+        if (e != null) {
+            e.printStackTrace(System.out);
+        }
+    }
+
+    @Override
+    public void run() {
+        if (Paxos._cloudLocked && !(H2O.CLOUD.leader() == H2O.SELF)) {
+            // cloud is formed the leader will take of subsequent refreshes
+            _executor.shutdown();
+            return;
+        }
+        for (int i = 0; i < _maxAttempts; i++) {
+            try {
+                Credentials creds = refreshTokens(loginAuthUser());
+                distribute(creds);
+                return;
+            } catch (IOException | InterruptedException e) {
+                log("Failed to refresh token (attempt " + i + " out of " + _maxAttempts + "). Will retry in " + _retryDelaySecs + "s.", e);
+            }
+            try {
+                Thread.sleep(_retryDelaySecs * 1000L);
+            } catch (InterruptedException e) {
+                Thread.currentThread().interrupt();
+            }
+        }
+    }
+
+    private Credentials refreshTokens(UserGroupInformation tokenUser) throws IOException, InterruptedException {
+        return tokenUser.doAs((PrivilegedExceptionAction<Credentials>) () -> {
+            Credentials creds = new Credentials();
+            Token<?>[] tokens = fetchDelegationTokens(_authUser, creds);
+            log("Fetched delegation tokens: " + Arrays.toString(tokens), null);
+            return creds;
+        });
+    }
+
+    private UserGroupInformation loginAuthUser() throws IOException, InterruptedException{
+        log("Log in from keytab as " + _authPrincipal, null);
+        UserGroupInformation realUser = UserGroupInformation.loginUserFromKeytabAndReturnUGI(_authPrincipal, _authKeytabPath);
```

Comment:
nit: the `realUser` variable seems to be redundant.

Reply:
yes, intellij agrees with you - however it is easier to debug it then (if I keep both variables)